### PR TITLE
Added support for pruning enum alternatives and struct fields.

### PR DESCRIPTION
**Pull request by gilescope**
(Apologies I seem to have cargo fmt the file at some point.)

Have a look at line 76 / 80 - seemed to be a bit more agressive if doesn't just do the top level.

These changes helped me reduce an ICE to something far more simpler. Your tool did all the commits marked 'reduce' https://github.com/gilescope/clique/commits/repro

Really impressed with the tools first outing. Hope to use it on many more ICEs.

### CODE REVIEW

**Comment by gilescope** [comment]
If your happy with the changes let me know and I'll add in some tests to keep the code company.

**Comment by jethrogb** [comment]
If you do the reformatting in a different commit, it's easier to review the actual changes.

File: src/transforms/prune_items.rs [hunk 1]
```diff
@@ -41,48 +42,100 @@ pub fn prune_items<F: FnMut(&syn::File) -> bool>(file: &mut syn::File, mut try_c
...
-			} else {
-				*index -= self.len();
-				false
-			}
-		} else {
-			false
-		}
-	}
+    fn prune(&mut self, level: usize, index: &mut usize) -> bool {
+        if level < 5 {
```
Comment (jethrogb) [inline src/transforms/prune_items.rs:80]: This change seems suspect, as well as the equivalent branches in the implementation below.

File: src/transforms/prune_items.rs [hunk 2]
```diff
@@ -41,48 +42,100 @@ pub fn prune_items<F: FnMut(&syn::File) -> bool>(file: &mut syn::File, mut try_c
...
+    }
 }
+
+impl Prune for syn::ItemEnum{
+    fn prune(&mut self, level: usize, index: &mut usize) -> bool {
+        if level < 5 {
+            if *index < self.variants.len() {
+                //  dbg!(my_enum.variants.clone());
+                let mut smaller : Punctuated<syn::Variant, syn::Token![,]> = Punctuated::new();
+                for (i, pair) in self.variants.pairs().enumerate() {
```
Comment (jethrogb) [inline src/transforms/prune_items.rs:100]: You can avoid the `clone` by using `variants.into_iter().enumerate()`. But you need to swap it out of `self` first. So the whole thing becomes:
```rust
let orig = std::mem::swap(&mut self.variants, Punctuated::new());
for (i, variant) in orig.into_iter().enumerate() {
    if i != *index {
        self.variants.push(variant);
    }
}
```

File: src/transforms/prune_items.rs [hunk 3]
```diff
@@ -41,48 +42,100 @@ pub fn prune_items<F: FnMut(&syn::File) -> bool>(file: &mut syn::File, mut try_c
...
+            false
+        }
+    }
 }
+
+impl Prune for syn::ItemEnum{
+    fn prune(&mut self, level: usize, index: &mut usize) -> bool {
+        if level < 5 {
+            if *index < self.variants.len() {
+                //  dbg!(my_enum.variants.clone());
```
Comment (jethrogb) [inline src/transforms/prune_items.rs:98]: Stray comment

File: src/transforms/prune_items.rs [hunk 4]
```diff
@@ -41,48 +42,100 @@ pub fn prune_items<F: FnMut(&syn::File) -> bool>(file: &mut syn::File, mut try_c
...
+                false
+            }
+        } else { false }
+    }
+}
+
+impl Prune for syn::ItemStruct {
+    fn prune(&mut self, level: usize, index: &mut usize) -> bool {
+        if level < 5 {
+            if let syn::Fields::Named(ref mut named) = self.fields {
```
Comment (jethrogb) [inline src/transforms/prune_items.rs:119]: Would be good to handle other types of structs as well.

**Pull request closed without merging**